Create inline-arrays.md (#7064)
dotnet/csharplang: proposals/safe-fixed-sized-buffers.md

Safe Fixed Size Buffers
=====

## Summary

> Provide a general-purpose and safe mechanism for declaring fixed sized buffers within C# classes, structs and interfaces.

Reviewer: For interfaces, it's specifically `static fields` only, is that correct?

Reviewer 2: I can assume that's implied since instance fields in interfaces aren't allowed (as far as the language is concerned, at least), but an explicit call-out would be good.

Author: > For interfaces, it's specifically `static fields` only, is that correct?

Yes, instance fields cannot be declared in an interface. I'll add an explicit note since they are not called fields in the speclet.

> Provide a general-purpose and safe mechanism for declaring fixed sized buffers within C# classes, structs and interfaces.

Reviewer 3: Nit: oxford comma.
```suggestion
Provide a general-purpose and safe mechanism for declaring fixed sized buffers within C# classes, structs, and interfaces.
```

## Motivation

This proposal plans to address the many limitations of https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/unsafe-code.md#228-fixed-size-buffers.

> Specifically it aims to allow the declaration of safe `fixed` buffers for managed and unmanaged types in a `struct`, `class`, or `interface`.
> And provide language safety verification for them.

Reviewer 3: The last line is a fragment, not a complete sentence.
```suggestion
Specifically it aims to allow the declaration of safe `fixed` buffers for managed and unmanaged types in a `struct`, `class`, or `interface`, and provide language safety verification for them.
```

## Detailed Design 

> Grammar for `variable_declarator` in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/classes.md#145-fields

Reviewer 3: ```suggestion
The grammar for *variable_declarator* in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/classes.md#145-fields
```

will be extended to allow specifying the size of the buffer:

``` antlr
field_declaration
    : attributes? field_modifier* type variable_declarators ';'
    ;

field_modifier
    : 'new'
    | 'public'
    | 'protected'
    | 'internal'
    | 'private'
    | 'static'
    | 'readonly'
    | 'volatile'
    | unsafe_modifier   // unsafe code support
    ;

variable_declarators
    : variable_declarator (',' variable_declarator)*
    ;
    
variable_declarator
    : identifier ('=' variable_initializer)?
    | fixed_size_buffer_declarator
    ;
    
fixed_size_buffer_declarator
    : identifier '[' constant_expression ']'
    ;    
```

> A `fixed_size_buffer_declarator` introduces a fixed-size buffers of a given element type.

Reviewer 3: ```suggestion
A *fixed_size_buffer_declarator* introduces a fixed-size buffers of a given element type.
```

> The buffer element type is the `type` specified in `field_declaration`. A fixed-size buffer declarator introduces a new member and consists of an identifier that names the member, followed by a constant expression enclosed in `[` and `]` tokens. The constant expression denotes the number of elements in the member introduced by that fixed-size buffer declarator. The type of the constant expression shall be implicitly convertible to type `int`, and the value shall be a non-zero positive integer.

Reviewer 3: ```suggestion
The buffer element type is the *type* specified in `field_declaration`. A fixed-size buffer declarator introduces a new member and consists of an identifier that names the member, followed by a constant expression enclosed in `[` and `]` tokens. The constant expression denotes the number of elements in the member introduced by that fixed-size buffer declarator. The type of the constant expression must be implicitly convertible to type `int`, and the value must be a non-zero positive integer.
```

> The buffer element type is the `type` specified in `field_declaration`. A fixed-size buffer declarator introduces a new member and consists of an identifier that names the member, followed by a constant expression enclosed in `[` and `]` tokens. The constant expression denotes the number of elements in the member introduced by that fixed-size buffer declarator. The type of the constant expression shall be implicitly convertible to type `int`, and the value shall be a non-zero positive integer.

Author: > I think that disallowing zero is fine, but did want to call out that C/C++ have "zero-length fixed sized buffers" and often use them as a way to indicate that the data is "variable length". 

No I don't want to call this out. The plan is to use https://github.com/dotnet/runtime/issues/61135 for metadata representation and it is said explicitly:
> `Length` must be greater than 0.

Apparently whatever C/C++ is using in not the same concept and is not using the same metadata representation.

The elements of a fixed-size buffer shall be laid out sequentially in memory as though they are elements of an array.

> A `volatile` modifier cannot be used with `fixed_size_buffer_declarator`. A System.ThreadStaticAttribute cannot be applied to a field declaration like that.

Reviewer 3: ```suggestion
A *field_declaration* with a *fixed_size_buffer_declarator* cannot have the `volatile` modifier, and it cannot be marked with a `System.ThreadStaticAttribute`.
```

> Depending on the situation (details are specified below), an access to a fixes-size buffer member is classified as a value (never a variable) of either

Reviewer 2: ```diff
- access to a fixes-size buffer
+ access to a fixed-size buffer
```

> `System.ReadOnlySpan<S>` or `System.Span<S>`, where S is the element type of the fixed-size buffer. Both types provide indexers returning refernce to a

Reviewer 2: Typo
```diff
- Both types provide indexers returning refernce
+ Both types provide indexers returning reference
```

> specific element with appropriate "readonly-ness". Which prevents direct assignment to the elements when language rules don't permit that.

Reviewer 2: ```diff
- appropriate "readonly-ness". Which prevents
+ appropriate "readonly-ness", which prevents
```

## Motivation

This proposal plans to address the many limitations of https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/unsafe-code.md#228-fixed-size-buffers.
Specifically it aims to allow the declaration of safe `fixed` buffers for managed and unmanaged types in a `struct`, `class`, or `interface`, and provide language safety verification for them.

## Detailed Design 

The grammar for *variable_declarator* in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/classes.md#145-fields
will be extended to allow specifying the size of the buffer:

``` antlr
field_declaration
    : attributes? field_modifier* type variable_declarators ';'
    ;

field_modifier
    : 'new'
    | 'public'
    | 'protected'
    | 'internal'
    | 'private'
    | 'static'
    | 'readonly'
    | 'volatile'
    | unsafe_modifier   // unsafe code support
    ;

variable_declarators
    : variable_declarator (',' variable_declarator)*
    ;
    
variable_declarator
    : identifier ('=' variable_initializer)?
    | fixed_size_buffer_declarator
    ;
    
fixed_size_buffer_declarator
    : identifier '[' constant_expression ']'
    ;    
```

> A *fixed_size_buffer_declarator* introduces a fixed-size buffers of a given element type.

Reviewer 4: >fixed-size buffers [](http://example.com/codeflow?start=46&length=18)

"fixed-size buffer" (singular)?

Author: Yes, it is a typo.

Indexing into the span with a constant expression outside of the declared fixed-size buffer bounds is a compile time error.

The *safe-to-escape* scope of the value will be equal to the *safe-to-escape* scope of the container, just as it would if the backing data was accessed as a field.

### Fixed-size buffers in expressions

Member lookup of a fixed-size buffer member proceeds exactly like member lookup of a field.

A fixed-size buffer can be referenced in an expression using a *simple_name* or a *member_access* .

When an instance fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `this.I`, where `I` is the fixed-size buffer member. When a static fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `E.I`, where `I` is the fixed-size buffer member and `E` is the declaring type.

#### Non-readonly fixed-size buffers

In a member access of the form `E.I`, if `E` is of a struct type and a member lookup of `I` in that struct type identifies an instance fixed-size member,
then `E.I` is evaluated and classified as follows:

> - If `E` is classified as a value, the result of the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of `I`.

Reviewer 4: >If `E` is classified as a value [](http://example.com/codeflow?start=2&length=31)

Does this cover the case of a non-readonly fixed-size field in a `readonly` method?

Author: > Does this cover the case of a non-readonly fixed-size field in a `readonly` method?

I think it does, `this` is qualified as a value rather than a variable in a `readonly` method, I think. In any case, my intent is to cover this scenario as well. If I discover that a special spec wording is necessary for that, I will add it.
